module/VuFind/src/VuFind/Marc/MarcReader.php

```diff
@@ -0,0 +1,264 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  MARC
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:record_drivers Wiki
+ */
+namespace VuFind\Marc;
+
+use VuFind\Marc\Serialization\Iso2709;
+use VuFind\Marc\Serialization\MarcXml;
+
+/**
+ * MARC record reader class.
+ *
+ * @category VuFind
+ * @package  MARC
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:record_drivers Wiki
+ */
+class MarcReader
+{
+    /**
+     * MARC is stored in a multidimensional array:
+     *  [001] - "12345"
+     *  [245] - i1: '0'
+     *          i2: '1'
+     *          s:  [
+     *                  ['a' => 'Title'],
+     *                  ['k' => 'Form'],
+     *                  ['k' => 'Another'],
+     *                  ['p' => 'Part'],
+     *              ]
+     */
+    protected $fields;
+
+    /**
+     * Constructor
+     *
+     * @param string $data MARC record in MARCXML or ISO2709 format
+     */
+    public function __construct($data)
+    {
+        $this->setData($data);
+    }
+
+    /**
+     * Set MARC record data
+     *
+     * @param string $data MARC record in MARCXML or ISO2709 format
+     *
+     * @throws Exception
+     * @return void
+     */
+    public function setData(string $data): void
+    {
+        if (MarcXml::canParse($data)) {
+            $this->fields = MarcXml::fromString($data);
+        } elseif (Iso2709::canParse($data)) {
+            $this->fields = Iso2709::fromString($data);
+        }
+    }
+
+    /**
+     * Serialize the record into MARCXML
+     *
+     * @return string
+     */
+    public function toXML(): string
+    {
+        return MarcXml::toString($this->getLeader(), $this->fields);
+    }
+
+    /**
+     * Convert to ISO2709.
+     *
+     * Returns an empty string if the record is too long.
+     *
+     * @return string
+     */
+    public function toISO2709(): string
+    {
+        return Iso2709::toString($this->getLeader(), $this->fields);
+    }
+
+    /**
+     * Return leader
+     *
+     * @return string
+     */
+    public function getLeader(): string
+    {
+        // Make sure leader is 24 characters. E.g. Voyager is often missing the last
+        // '0' of the leader.
+        return isset($this->fields['000'])
+            ? str_pad(substr($this->fields['000'], 0, 24), 24) : '';
```

Comment:
Would it be better to do this normalization logic when setting data in the first place, so that it only needs to be applied once even if the leader is requested multiple times? Or is it actually more efficient to leave it as-is and only do it on demand?

Reply:
Yes. I moved the leader out of fields altogether, which made the handling a bit cleaner too.